perf/src/recycled_vec.rs

```diff
@@ -0,0 +1,221 @@
+    type Iter = rayon::slice::IterMut<'a, T>;
+    type Item = &'a mut T;
+    fn into_par_iter(self) -> Self::Iter {
+        self.x.par_iter_mut()
+    }
+}
+
+impl<T: Clone + Default + Sized> RecycledVec<T> {
+    pub fn reserve(&mut self, size: usize) {
+        self.x.reserve(size);
+    }
+
+    pub fn copy_from_slice(&mut self, data: &[T])
+    where
+        T: Copy,
+    {
+        self.x.copy_from_slice(data);
+    }
+
+    pub fn from_vec(source: Vec<T>) -> Self {
+        Self {
+            x: source,
+            recycler: Weak::default(),
+        }
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self::from_vec(Vec::with_capacity(capacity))
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.x.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.x.len()
+    }
+
+    pub fn as_ptr(&self) -> *const T {
+        self.x.as_ptr()
+    }
+
+    pub fn as_mut_ptr(&mut self) -> *mut T {
+        self.x.as_mut_ptr()
+    }
+
+    pub fn push(&mut self, x: T) {
+        self.x.push(x);
+    }
+
+    pub fn truncate(&mut self, size: usize) {
+        self.x.truncate(size);
+    }
+
+    pub fn resize(&mut self, size: usize, elem: T) {
+        self.x.resize(size, elem);
+    }
+
+    pub fn append(&mut self, other: &mut Vec<T>) {
+        self.x.append(other);
+    }
+
+    pub fn append_pinned(&mut self, other: &mut Self) {
+        self.x.append(&mut other.x);
+    }
+
+    /// Forces the length of the vector to `new_len`.
+    ///
+    /// This is a low-level operation that maintains none of the normal
+    /// invariants of the type. Normally changing the length of a vector
+    /// is done using one of the safe operations instead, such as
+    /// [`truncate`], [`resize`], [`extend`], or [`clear`].
+    ///
+    /// [`truncate`]: Vec::truncate
+    /// [`resize`]: Vec::resize
+    /// [`extend`]: Extend::extend
+    /// [`clear`]: Vec::clear
+    ///
+    /// # Safety
+    ///
+    /// - `new_len` must be less than or equal to [`capacity()`].
+    /// - The elements at `old_len..new_len` must be initialized.
+    ///
+    /// [`capacity()`]: Vec::capacity
+    ///
+    pub unsafe fn set_len(&mut self, size: usize) {
+        self.x.set_len(size);
+    }
+
+    pub fn shuffle<R: Rng>(&mut self, rng: &mut R) {
+        self.x.shuffle(rng)
+    }
+}
+
+impl<T: Clone + Default + Sized> Clone for RecycledVec<T> {
+    fn clone(&self) -> Self {
+        let x = self.x.clone();
+        debug!("clone RecycledVec: size: {}", self.x.capacity(),);
+        Self {
+            x,
+            recycler: self.recycler.clone(),
+        }
+    }
+}
+
+impl<T: Sized + Default + Clone> Drop for RecycledVec<T> {
+    fn drop(&mut self) {
+        if let Some(recycler) = self.recycler.upgrade() {
+            recycler.recycle(std::mem::take(self));
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_pinned_vec() {
```

Comment:
```suggestion
    fn test_recycled_vec() {
```

Reply:
@apfitzge This is not addressed yet :slightly_smiling_face:

Reply:
ac660edf75f449b6594fa1d502be6477e41bbbb6